### ISSUE TITLE
flash.system.Capabilities.language returns zh-CN instead of zh-TW for Chinese, Traditional on iOS

### ISSUE THREAD

**Issue by jackwlee01** [Gamua/Adobe-Runtime-Support#349]
On iPhone 6+ 12.4.2 and iPhone X 13.5.1, using Air 33.1.

In iOS Settings I change my language to Chinese, Traditional.
flash.system.Capabilities.language returns zh-CN instead of zh-TW.

Note that if I set my language to Chinese, Traditional (Hong Kong) flash.system.Capabilities.language appropriately returns zh-TW.

**Comment by JohnBlackburne** [comment]
This is definitely a bug [flash.system.Capabilities.language](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/system/Capabilities.html#language) has `zh-CN` for simplified Chinese, `zh-TW` for traditional Chinese.

But it's confusingly named as the proper language code for simplified Chinese is `zh-Hans`, for traditional Chinese `zh-Hant`. Suffixes `-CN` and `-TW` can be used to further specify the country; see [Language Tags](https://www.iana.org/assignments/language-tags/language-tags.xml). That is perhaps where the confusion arose. I note that [iOS language codes](https://gist.github.com/xta/6e9b63db1fa662bb3910b680f9ebd458) match the proper language codes too.

**Comment by ajwfrost** [comment]
Hi

Yes.. the iOS code for Traditional Chinese without a country specifier is coming to us as “zh-hant-US” and its this “US” which causes the AIR logic to fail. We have a fix for it though..

In terms of switching to using zh-Hans and zh-Hant, I agree this is more proper but am concerned we’d see this breaking some apps that rely on the current behaviour... Would be interested by folks’ feedback on this sort of thing!

Thanks

**Comment by JohnBlackburne** [comment]
I would not recommend switching to `zh-Hant`, `zh-Hans`. I am a bit of a language geek so know the proper codes, but what's there now makes sense as long as it works as documented. I think `zh-CN` and `zh-TW` corresponds to  the `/Resources` sub-folder names in an app, so makes sense from that point of view too.

**Comment by zwetan** [comment]
> In terms of switching to using zh-Hans and zh-Hant, I agree this is more proper but am concerned we’d see this breaking some apps that rely on the current behaviour... Would be interested by folks’ feedback on this sort of thing!

see [Move zh-tw and zh-cn locales to zh-hant and zh-hans, respectively](https://github.com/mozilla/addons-server/issues/5829)
and [Full backwards compatibility for old language codes](https://github.com/django/django/commit/e5e044da87800feb6ef63fef1765d8c05022d926)

imho `zh-hant` and `zh-hans` should be added while still supporting `zh-tw` and `zh-cn` for backward compatibility

**Comment by JohnBlackburne** [comment]
I don't think you could add `zh-Hant` and `zh-Hans` without breaking existing code which checks for and expects `zh-TW` and `zh-CN`.

And I think you can get correct and more detailed locale information from [flash.system.Capabilities.languages](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/system/Capabilities.html#languages). In particular the first element of the array returned by `languages` should be the OS supplied version of the user's preferred locale, so `zh-Hant-HK` or whatever it is.
